pbrd/pbr_nht.c

```diff
@@ -0,0 +1,699 @@
+{
+	;
+}
+
+void pbr_nht_route_removed_for_table(uint32_t table_id)
+{
+	hash_iterate(pbr_nhg_hash, pbr_nht_find_nhg_from_table_remove,
+		     &table_id);
+}
+
+/*
+ * Loop through all nexthops in a nexthop group to check that they are all the
+ * same. If they are not all the same, log this peculiarity.
+ *
+ * nhg
+ *    The nexthop group to check
+ *
+ * Returns:
+ *    - AFI of last nexthop in the group
+ *    - AFI_MAX on error
+ */
+static afi_t pbr_nht_which_afi(struct nexthop_group nhg)
+{
+	struct nexthop *nexthop;
+	afi_t install_afi = AFI_MAX;
+	bool v6, v4, bh;
+	v6 = v4 = bh = false;
+
+	for (ALL_NEXTHOPS(nhg, nexthop)) {
+		switch (nexthop->type) {
+		case NEXTHOP_TYPE_IFINDEX:
+			break;
+		case NEXTHOP_TYPE_IPV4:
+		case NEXTHOP_TYPE_IPV4_IFINDEX:
+			v6 = true;
+			install_afi = AFI_IP;
+			break;
+		case NEXTHOP_TYPE_IPV6:
+		case NEXTHOP_TYPE_IPV6_IFINDEX:
+			v4 = true;
+			install_afi = AFI_IP6;
+			break;
+		case NEXTHOP_TYPE_BLACKHOLE:
+			bh = true;
+			install_afi = AFI_MAX;
+			break;
+		}
+	}
+
+	if (!bh && v6 && v4)
+		DEBUGD(&pbr_dbg_nht,
+		       "%s: Saw both V6 and V4 nexthops...using %s",
+		       __PRETTY_FUNCTION__, afi2str(install_afi));
+	if (bh && (v6 || v4))
+		DEBUGD(&pbr_dbg_nht,
+		       "%s: Saw blackhole nexthop(s) with %s%s%s nexthop(s), using AFI_MAX.",
+		       __PRETTY_FUNCTION__, v4 ? "v4" : "",
+		       (v4 && v6) ? " and " : "", v6 ? "v6" : "");
+
+	return install_afi;
+}
+
+static void pbr_nht_install_nexthop_group(struct pbr_nexthop_group_cache *pnhgc,
+					  struct nexthop_group nhg)
+{
+	afi_t install_afi;
+
+	install_afi = pbr_nht_which_afi(nhg);
+
+	pnhgc->installed = false;
+	route_add(pnhgc, nhg, install_afi);
+}
+
+static void
+pbr_nht_uninstall_nexthop_group(struct pbr_nexthop_group_cache *pnhgc,
+				struct nexthop_group nhg)
+{
+	afi_t install_afi;
+
+	install_afi = pbr_nht_which_afi(nhg);
+
+	pnhgc->installed = false;
+	pnhgc->valid = false;
+	route_delete(pnhgc, install_afi);
+}
+
+void pbr_nht_change_group(const char *name)
+{
+	struct nexthop_group_cmd *nhgc;
+	struct pbr_nexthop_group_cache *pnhgc;
+	struct pbr_nexthop_group_cache find;
+	struct nexthop *nhop;
+
+	nhgc = nhgc_find(name);
+	if (!nhgc)
+		return;
+
+	memset(&find, 0, sizeof(find));
+	strcpy(find.name, name);
+	pnhgc = hash_lookup(pbr_nhg_hash, &find);
+
+	if (!pnhgc) {
+		DEBUGD(&pbr_dbg_nht,
+		       "%s: Could not find nexthop-group cache w/ name '%s'",
+		       __PRETTY_FUNCTION__, name);
+		return;
+	}
+
+	for (ALL_NEXTHOPS(nhgc->nhg, nhop)) {
+		struct pbr_nexthop_cache lookup;
+		struct pbr_nexthop_cache *pnhc;
+
+		memcpy(&lookup.nexthop, nhop, sizeof(*nhop));
+		pnhc = hash_lookup(pnhgc->nhh, &lookup);
+		if (!pnhc) {
+			pnhc = hash_get(pnhgc->nhh, &lookup, pbr_nh_alloc);
+			pnhc->parent = pnhgc;
+		}
+	}
+	pbr_nht_install_nexthop_group(pnhgc, nhgc->nhg);
+}
+
+char *pbr_nht_nexthop_make_name(char *name, uint32_t seqno, char *buffer)
+{
+	sprintf(buffer, "%s%u", name, seqno);
+	return buffer;
```

Comment:
we should use snprintf, instead of sprintf.

Reply:
fixed

Reply:
thanks